plugin/src/main/java/com/playmonumenta/scriptedquests/commands/Music.java

```diff
@@ -0,0 +1,298 @@
+import javax.annotation.Nullable;
+import org.bukkit.Bukkit;
+import org.bukkit.NamespacedKey;
+import org.bukkit.SoundCategory;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Player;
+import org.bukkit.scheduler.BukkitRunnable;
+
+public class Music {
+	private static class Song {
+		public String mSongPath;
+		public SoundCategory mCategory;
+		public long mSongDuration;
+		public boolean mIsLoop;
+		public float mVolume;
+		public float mPitch;
+
+		public Song(String path, SoundCategory category, double durationSeconds, boolean isLoop, float volume, float pitch) {
+			mSongPath = path;
+			mCategory = category;
+			mSongDuration = (long) (1000.0f * durationSeconds);
+			mIsLoop = isLoop;
+			mVolume = volume;
+			mPitch = pitch;
+		}
+
+		public boolean equalsSong(Song song) {
+			return mIsLoop == song.mIsLoop
+				&& mSongDuration == song.mSongDuration
+				&& mVolume == song.mVolume
+				&& mPitch == song.mPitch
+				&& mCategory.equals(song.mCategory)
+				&& mSongPath.equals(song.mSongPath);
+		}
+	}
+
+	private static class PlayerState implements Runnable {
+		private static final ZoneId TIMEZONE = ZoneOffset.UTC;
+
+		private final UUID mPlayerId;
+		private @Nullable Song mNow = null;
+		private @Nullable Song mNext = null;
+		private LocalDateTime mNextTime = LocalDateTime.MAX;
+
+		public PlayerState(UUID playerId) {
+			mPlayerId = playerId;
+		}
+
+		public void playNow() {
+			Player player = Bukkit.getPlayer(mPlayerId);
+			if (player == null) {
+				cancelNext();
+				cancelNow();
+				return;
+			}
+
+			if (millisToRefresh() <= 0) {
+				cancelNow();
+			}
+
+			if (mNow == null) {
+				if (mNext == null) {
+					return;
+				}
+				mNow = mNext;
+				if (!mNext.mIsLoop) {
+					mNext = null;
+				}
+			}
+
+			mNextTime = LocalDateTime.now(TIMEZONE).plus(mNow.mSongDuration, ChronoUnit.MILLIS);
+			player.playSound(player.getLocation(), mNow.mSongPath, mNow.mCategory, mNow.mVolume, mNow.mPitch);
+			mRealTimePool.schedule(this, millisToRefresh(), TimeUnit.of(ChronoUnit.MILLIS));
+		}
+
+		public void playNow(Song song) {
+			cancelNext();
+			if (mNow != null && !mNow.equalsSong(song)) {
+				cancelNow();
+			}
+			playNext(song);
+		}
+
+		public void cancelNow() {
+			if (mNow == null) {
+				return;
+			}
+
+			Player player = Bukkit.getPlayer(mPlayerId);
+			if (player != null) {
+				player.stopSound(mNow.mSongPath, mNow.mCategory);
+			}
+			mNow = null;
+			mNextTime = LocalDateTime.MAX;
+		}
+
+		public void playNext(Song song) {
+			mNext = song;
+			if (mNow == null || millisToRefresh() <= 0) {
+				playNow();
+			}
+		}
+
+		public void cancelNext() {
+			mNext = null;
+		}
+
+		public long millisToRefresh() {
+			try {
+				return LocalDateTime.now(TIMEZONE).until(mNextTime, ChronoUnit.MILLIS);
+			} catch (ArithmeticException ignored) {
+				return Long.MAX_VALUE;
+			}
+		}
+
+		@Override
+		public void run() {
+			if (millisToRefresh() <= 0) {
+				new BukkitRunnable() {
```

Comment:
could be simplified to Bukkit.getScheduler().runTask(Plugin.getInstance(), this::playNow);

Reply:
Apparently not:
> Reference to 'playNow' is ambiguous, both 'playNow()' and 'playNow(Song)' match

Reply:
just () -> playNow() then
Strange that it wouldn't know you're referring to the one with no arguments as the other doesn't make sense